11/umbraco-cms/reference/searching/examine/indexing.md

```diff
@@ -305,29 +304,43 @@ namespace Umbraco.Docs.Samples.Web.CustomIndexing
             _productIndexValueSetBuilder = productIndexValueSetBuilder;
             RegisterIndex("ProductIndex");
         }
+
         protected override void PopulateIndexes(IReadOnlyList<IIndex> indexes)
         {
-            foreach (var index in indexes)
+            foreach (IIndex index in indexes)
             {
-                var roots = _contentService.GetRootContent();
-
-                index.IndexItems(_productIndexValueSetBuilder.GetValueSets(roots.ToArray()));
+                IContent[] roots = _contentService.GetRootContent().ToArray();
+                index.IndexItems(_productIndexValueSetBuilder.GetValueSets(roots));
 
-                foreach (var root in roots)
+                foreach (IContent root in roots)
                 {
-                    var valueSets = _productIndexValueSetBuilder.GetValueSets(_contentService.GetPagedDescendants(root.Id, 0, Int32.MaxValue, out _).ToArray());
-                    index.IndexItems(valueSets);
+                    const int pageSize = 10000;
+                    var pageIndex = 0;
+                    IContent[] descendants;
+                    do
+                    {
+                        descendants = _contentService.GetPagedDescendants(root.Id, pageIndex, pageSize, out _).ToArray();
+                        IEnumerable<ValueSet> valueSets = _productIndexValueSetBuilder.GetValueSets(descendants);
+                        index.IndexItems(valueSets);
+
+                        pageIndex++;
+                    }
+                    while (descendants.Length == pageSize);
                 }
             }
-
         }
     }
 }
 ```
+{% hint style="info" %}
+Note here that this is just an example of how you could do indexing, this means that we're indexing all content, also unpublished & trashed content.
```

Comment:
```suggestion
Note that this is just an example of how you could do indexing. In this sample we're indexing all content, also unpublished content.
```

Since we're only fetching descendants of content roots, we can't ever be indexing trashed content here?